src/ncp/ncp_base.hpp

```diff
@@ -780,6 +805,34 @@ class NcpBase
 #endif
 
 #if OPENTHREAD_CONFIG_NCP_DNSSD_ENABLE && OPENTHREAD_CONFIG_PLATFORM_DNSSD_ENABLE
+
+    template <typename DnssdObjType>
+    void DnssdUpdate(const DnssdObjType         *Obj,
+                     otPlatDnssdRequestId        aRequestId,
+                     otPlatDnssdRegisterCallback aCallback,
+                     bool                        aRegister)
+    {
+        otError          error  = OT_ERROR_NONE;
+        uint8_t          header = SPINEL_HEADER_FLAG | SPINEL_HEADER_TX_NOTIFICATION_IID;
+        spinel_command_t cmd    = aRegister ? SPINEL_CMD_PROP_VALUE_INSERTED : SPINEL_CMD_PROP_VALUE_REMOVED;
+
+        VerifyOrExit(mDnssdState == OT_PLAT_DNSSD_READY, error = OT_ERROR_INVALID_STATE);
+
+        SuccessOrExit(error = mEncoder.BeginFrame(header, cmd));
+        SuccessOrExit(error = EncodeDnssd(aRequestId, Obj));
+        SuccessOrExit(error = mEncoder.WriteData(reinterpret_cast<const uint8_t *>(&aCallback), sizeof(aCallback)));
```

Comment:
One suggestion here: We can consider reordering the fields in the spinel properties to place `aRequestId` and `aCallback` (as data) next to each other and after all other fields?

This way, `EncodeDnssd<>` could encode `Obj` only (no need for it to get `aRequestid`, and then the request ID and `aCallback` could be added as common code for all `<ObjType>` templates in this method.